### DISCUSSION TITLE
getChannel and removeAllChannel() not working?

### DISCUSSION THREAD

**Discussion by pierroo** [supabase/supabase#20610] [Questions] [answered]
I am using supabase in React Native.

I subscribe to a channel from screen A
![image](https://github.com/supabase/supabase/assets/17619412/ce10ba63-972d-49ba-9940-bc5aa2aab388)

I later use supabase.getChannel() just for the sake of testing, and it displays an empty array
=> getChannel not working?

Another example, upon moving to screen B I call a cleanup function to unsubscribe from all channel (specifically the previously mentioned, but just for the sake of testing) so I call supabase.removeAllChannels()
But then if I go back and forth many times from A to B etc, I can see that those subscribers to channel are stacking up (from console logs on different events)
=> removeAllChannel() not working?

I know there are opened channels since I can see events related to them.
And I know that if I remove a specific channel from its previously created reference it does work.

It's just the getChannel and removeAllChannel that seems to not be working?

On the other hand, does it matter from which supabase client I call those function?
I use two ways to instantiate my supabase createClient, one with a token in header, one without using the super basic createClient

Will getChannel / removeChannel / removeAllChannel return the same whether I'm using supabaseNotAuthed and supabaseAuthed?
(both pointing to the same client keys etc, just without/without token in header)

![image](https://github.com/supabase/supabase/assets/17619412/8261e59b-3e25-4bfb-a71f-90078426051e)

**Comment by GaryAustin1** [comment]
All realtime activities should be on the same client.  The client contains all state and objects.  Other clients will not have access to this state.

**Reply by pierroo** [reply]
Thank you for your answer @GaryAustin1 , unfortunately I'm not sure I was clear; I will try to clarify:

Both clients creation are pointing toward exactly the same supabase:
![image](https://github.com/supabase/supabase/assets/17619412/cc85e3e1-74da-49da-89f6-5deb6bd94438)

It's just that sometimes I only need supabase.function()
and sometimes since I rely on RLS I pass a token and use supabaseAuthed(token).function()

Both are pointing toward the same.

Or do you mean that each time I call the supabase createClient it will somehow create a new reference that just accumulates in parallel without them ever being the same?

For example everytime I send a chat message I re-call supabase.createClient with token:
![image](https://github.com/supabase/supabase/assets/17619412/279c9549-8933-4488-af52-6c6bac4c782d)

Is this not the way to do?

**Reply by GaryAustin1** [reply, marked as answer]
createClient creates an object in memory called in your case supabase and supabaseAuthed.  You can add channels to realtime to that client and remove them.  If you call createClient again using supabase = createClient() then the old supabase client is gone if running on a browser and all realtime channels associated with it or gone.  All state is maintained in the object created by createClient.

**Reply by pierroo** [reply]
"If you call createClient again using supabase = createClient() then the old supabase client is gone if running on a browser" 
=> is this also true in react native / mobile app?

so if I were to use supabaseClient ONLY (no "naked supabase without token") following the supabaseAuthed each and everytime I send a chat message (even multiple times in a row) there would be no impact / performance issue / and that's how it should be done?

**Reply by GaryAustin1** [reply]
Normally you create a client once on a browser and use it until the tab closes.   This maintains all session/realtime state. 

If you start it over each time you don't do a server call but you lose your realtime connections and you load the session from local storage each time so the user session should still be maintained.  This is not the normal operation on a browser.

Server side is different as you normally DO want a new client each time as the server could be handling lots of users.
